src/storage/btree_load.c

```diff
@@ -3878,3 +3750,457 @@ btree_get_perf_btree_page_type (THREAD_ENTRY * thread_p, PAGE_PTR page_ptr)
+							     is_fk_scan_desc, &pk_key_cnt, &(pk_header), NULL);
+
+	      if (ret != NO_ERROR)
+		{
+		  goto end;
+		}
+
+	      /* We need to compare the current value with the new value from the primary key. */
+	      compare_ret = btree_compare_key (&pk_key, &fk_key, pk_bt_scan.btid_int.key_type, 1, 1, NULL);
+
+	      if (compare_ret == DB_EQ)
+		{
+		  /* Found value, stop searching in pk. */
+		  break;
+		}
+	      else if (compare_ret == DB_LT)
+		{
+		  /* No match yet. Advance in pk. */
+		  pk_bt_scan.slot_id += 1;
+		  continue;
+		}
+	      else
+		{
+		  /* Fk is invalid. Set error. */
+		  val_print = pr_valstring (&fk_key);
+		  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_FK_INVALID, 2, sort_args->fk_name,
+			  (val_print ? val_print : "unknown value"));
+		  ret = ER_FK_INVALID;
+		  goto end;
+		}
+	    }
+	  /* Check whether the key no longer resides in the current page. If so, restart scan from top. */
+	  if (!found && pk_bt_scan.slot_id > pk_key_cnt)
+	    {
+	      old_page = pk_bt_scan.C_page;
+	      pk_bt_scan.C_page = NULL;
+	      goto start_primary_search;
+	    }
+	}
+
+      if (found == true)
+	{
+	  pr_clear_value (&fk_key);
+	}
+
+      pr_clear_value (&pk_key);
+
+    }
+end:
+
+  if (old_page != NULL)
+    {
+      pgbuf_unfix_and_init (thread_p, old_page);
+    }
+
+  if (curr_fk_pageptr != NULL)
+    {
+      pgbuf_unfix_and_init (thread_p, curr_fk_pageptr);
+    }
+
+  if (pk_bt_scan.C_page)
+    {
+      pgbuf_unfix_and_init (thread_p, pk_bt_scan.C_page);
+    }
+
+  if (load_args->leaf.pgptr)
+    {
+      pgbuf_unfix (thread_p, load_args->leaf.pgptr);
+    }
+
+  if (!DB_IS_NULL (&fk_key))
+    {
+      pr_clear_value (&fk_key);
+    }
+  if (!DB_IS_NULL (&pk_key))
+    {
+      pr_clear_value (&pk_key);
+    }
+
+  return ret;
+}
+
+/*
+ * btree_get_value_from_leaf_slot () -
+ *
+ *   return:			  - NO_ERROR or error code.
+ *   btid_int(in):		  - The structure of the B-tree where the leaf resides.
+ *   leaf_ptr(in):		  - The leaf where the value needs to be extracted from.
+ *   slot_id(in):		  - The slot from where the value must be pulled.
+ *   key(out):			  - The value requested.
+ *
+ */
+static int
+btree_get_value_from_leaf_slot (THREAD_ENTRY * thread_p, BTID_INT * btid_int, PAGE_PTR leaf_ptr, int slot_id,
+				DB_VALUE * key)
+{
+  LEAF_REC leaf;
+  bool clear_first_key = false;
+  int first_key_offset = 0;
+  RECDES record;
+  int ret = NO_ERROR;
+
+  if (spage_get_record (thread_p, leaf_ptr, slot_id, &record, PEEK) != S_SUCCESS)
+    {
+      assert_release (false);
+      ret = ER_FAILED;
+      return ret;
+    }
+
+  ret =
+    btree_read_record (thread_p, btid_int, leaf_ptr, &record, key,
+		       &leaf, BTREE_LEAF_NODE, &clear_first_key, &first_key_offset, PEEK_KEY_VALUE, NULL);
+
+  if (ret != NO_ERROR)
+    {
+      ASSERT_ERROR ();
+      return ret;
+    }
+
+  return ret;
+}
+
+/*
+ * btree_advance_to_next_slot_and_fix_page () -
```

Comment:
Add function description mentioning that it advances from one page to another without conditional latches. Warn callers that it should not used for descending scans if no concurrency cannot be guaranteed.